staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/object_count_tracker.go

```diff
@@ -0,0 +1,133 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+)
+
+// StorageObjectCountTracker is an interface that is used to keep track of
+// of the total number of objects for each resource.
+// {group}.{resource} is used as the key name to update and retrieve
+// the total number of objects for a given resource.
+type StorageObjectCountTracker interface {
+	// OnCount is invoked to update the current number of total
+	// objects for the given resource
+	OnCount(string, int64)
+
+	// Get returns the total number of objects for the given resource.
+	// If the given resource is not being tracked Get will return zero.
+	// For now, we do not differentiate between zero object count and
+	// a given resoure not being present.
+	Get(string) int64
+}
+
+// NewStorageObjectCountTracker returns an instance of
+// StorageObjectCountTracker interface that can be used to
+// keep track of the total number of objects for each resource.
+func NewStorageObjectCountTracker(stopCh <-chan struct{}) StorageObjectCountTracker {
+	const (
+		// type deletion (it applies mostly to CRD) is not a very frequent operation
+		// so we can afford to prune the cache at a large interval.
+		pruneInterval = 12 * time.Hour
+	)
+	tracker := &objectCountTracker{
+		clock:  &clock.RealClock{},
+		counts: map[string]*timestampedCount{},
+	}
+	go func() {
+		wait.PollUntil(
+			pruneInterval,
+			func() (bool, error) {
+				tracker.prune(pruneInterval)
+				return false, nil
+			}, stopCh)
+		klog.InfoS("StorageObjectCountTracker pruner is exiting")
+	}()
+
+	return tracker
+}
+
+// timestampedCount stores the count of a given resource with a last updated
+// timestamp so we can prune it after it goes stale for certain threshold.
+type timestampedCount struct {
+	count       int64
+	lastUpdated time.Time
+}
+
+// objectCountTracker implements StorageObjectCountTracker with
+// reader/writer mutual exclusion lock.
+type objectCountTracker struct {
+	clock clock.PassiveClock
+
+	lock   sync.RWMutex
+	counts map[string]*timestampedCount
+}
+
+func (t *objectCountTracker) OnCount(groupResource string, count int64) {
+	if count == -1 {
+		// a value of -1 indicates that the 'Count' call failed to contact
+		// the storage layer, in most cases this error can be transient.
+		// we will continue to work with the count that is in the cache
+		return
+	}
+
+	now := t.clock.Now()
+
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if item, ok := t.counts[groupResource]; ok {
+		item.count = count
+		item.lastUpdated = now
+		return
+	}
+
+	t.counts[groupResource] = &timestampedCount{
+		count:       count,
+		lastUpdated: now,
+	}
+}
+
+func (t *objectCountTracker) Get(groupResource string) int64 {
```

Comment:
As mentioned above (and also as discussed in the comment thread in the other PR), we actually should be distinguishing 0 from "unknown type" (and also from "stale data").

So I suggest returning (int64, error), where error can be one of:
- unknown type
- stale data
and the called can handle them appropriately

Reply:
I am not sure if I have the context to know whether it's an "unknown type" from inside `Store.startObservingCount`. As far as APF is concerned, we can work with stale data for at most an hour now. So when we get `(0, false)` from Get we can treat this as count unknown/unknown type, thoughts?